src/compiler/dataStructures.ts

```diff
@@ -0,0 +1,768 @@
+    }
+
+    /** Number-keyed Map. */
+    export interface NMap<V> extends MapCommon<number, V> {}
+    export interface NMapConstructor {
+        new(): NMap<any>;
+        new<V>(entries?: [number, V][]): NMap<V>;
+    }
+
+    abstract class ShimMapCommon<K extends any, V> implements MapCommon<K, V> {
+        protected data: any; // ObjMap<V> | V[]
+
+        abstract clear(): void;
+
+        delete(key: K) {
+            delete this.data[<any>key];
+        }
+
+        get(key: K) {
+            return this.has(key) ? this.data[<any>key] : undefined;
+        }
+
+        has(key: K) {
+            return hasProperty(this.data, <any>key);
+        }
+
+        set(key: K, value: V) {
+            this.data[<any>key] = value;
+        }
+    }
+
+    /** Implementation of Map for JS engines that don't natively support it. */
+    class ShimSMap<V> extends ShimMapCommon<string, V> implements SMap<V> {
+        data: ObjMap<V>;
+
+        constructor(entries?: [string, V][]);
+        constructor(otherMap: SMap<V>);
+        constructor(argument?: any) {
+            super();
+            this.data = {};
+
+            if (argument !== undefined) {
+                if (argument instanceof ShimSMap) {
+                    this.data = clone(argument.data);
+                }
+                else {
+                    Debug.assert(argument instanceof Array);
+                    for (const [key, value] of argument) {
+                        this.set(key, value);
+                    }
+                }
+            }
+        }
+
+        clear() {
+            this.data = {};
+        }
+
+        forEach(fn: (value: V, index: string) => void) {
+            for (const key in this.data) {
+                if (this.has(key)) {
+                    fn(this.data[key], key);
+                }
+            }
+        }
+
+        // Support this directly instead of emulating an iterator, which is slow.
+        find<U>(fn: (value: V, key: string) => U | undefined): U | undefined {
+            for (const key in this.data) {
+                if (this.has(key)) {
+                    const result = fn(this.data[key], key);
+                    if (result) {
+                        return result;
+                    }
+                }
+            }
+        }
+    }
+
+    /** Implementation of Map for JS engines that don't natively support it. */
+    class ShimNMap<V> extends ShimMapCommon<number, V> implements NMap<V> {
+        data: V[];
+
+        constructor(entries?: [number, V][]) {
+            super();
+            this.data = [];
+            if (entries) {
+                for (const [key, value] of entries) {
+                    this.set(key, value);
+                }
+            }
+        }
+
+        clear() {
+            this.data = [];
+        }
+    }
+
+    declare const Map: (SMapConstructor & NMapConstructor) | undefined;
+    export const SMap: SMapConstructor = Map ? Map : ShimSMap;

```

Comment:
Question: since `Map` is not defined here then it will always use `ShimSMap`?  I feel like I may miss something...